DEVREL-977 - Flatten circular data, fix product filtering

pokornyd opened the pull request:
### Motivation

[DEVREL-977](https://kontent-ai.atlassian.net/jira/software/projects/DEVREL/boards/421?selectedIssue=DEVREL-977)

- replacement PR for https://github.com/kontent-ai/sample-app-next-js/pull/56

Adds helper methods to flatten/restore items with potential circular references to pass through getStaticProps. Fixes a bug where product filtering wasn't applied at all.

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

Introduce a circular reference in Kontent.ai project (e.g. via content item link in any given Fact item), run the app and visit any page that requests said item, confirm it works as expected.

IvanKiral commented on pages/[envId]/[slug].tsx:
```diff
@@ -48,41 +49,48 @@ export const getStaticProps: GetStaticProps<Props, IParams> = async (context) =>
     };
   }
 
+  const page = stringifyAsType(pageData);
```
I don't understand this. Why do I only send a string with the type definition to the client, and then let him parse data back to the object? isn't it easier to do the stringify and parsing here at the backend and send the client just what it needs? Or am I missing something?

pokornyd replied:
getStaticProps serializes the values you return in it. if they have circular references, the app will crash. instead, we first flatten the object into a string format, then restore its data client-side, thus avoiding any circular issues in getStaticProps without losing any information from the object.

IvanKiral commented on lib/utils/changeUrlQueryString.ts:
```diff
@@ -2,5 +2,10 @@ import { NextRouter } from "next/router";
 export const changeUrlQueryString = (query: ParsedUrlQueryInput, router: NextRouter) => {
-  router.replace({ query: query }, undefined, { scroll: false, shallow: true });
+  const { envId, ...restQuery } = query; // get rid of envId
+  const asPath = {
```
this comment doesn't say to me much why are getting rid of envid? better comment would explain why do we need asPath without the envid? is it because it is handled component and won't by the middleware?

pokornyd replied:
If I recall, the idea with products was that they'll be filtered client-side, unlike articles that are pulled from API. that was in pre-middleware times though, so I'm open to suggestions. In any case, the spread operator is used to get rid of `envId` in construction of the subsequent `asPath`, used in router.replace to ensure the envId is not present in the address bar.

IvanKiral reviewed (changes requested):
Now when use both categories and paging, the URL search bar only has one of the query params (categories or pagin).. also we should consider writing a readme section about the handling circular references

pokornyd commented:
> Now when use both categories and paging, the URL search bar only has one of the query params (categories or pagin).. also we should consider writing a readme section about the handling circular references
> 

I refactored the method and added comments to make the steps clear. Also added a readme entry.
